File: api/schemas/task.py
"""Schemas for Task model in the Spirited Todo List API."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field

from models.task import Priority


class TaskBase(BaseModel):
    """Base schema for Task model."""

    title: str = Field(..., min_length=1)
    description: Optional[str] = None
    priority: Priority = Priority.LOW
    deadline: Optional[datetime] = None


class TaskCreate(TaskBase):
    """Schema for creating a new Task."""


class TaskRead(TaskBase):
    """Schema for reading a Task."""

    id: int
    created_at: datetime
    updated_at: datetime


class TaskUpdate(BaseModel):
    """Schema for updating a Task."""

    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[Priority] = None
    deadline: Optional[datetime] = None


class TaskListResponse(BaseModel):
    """Schema for listing tasks."""

    items: List[TaskRead]
    total: int
    page: int
    page_size: int

File: web/src/types/task.ts
export enum Priority {
  LOW = 1,
  MID = 2,
  HIGH = 3,
}

export interface Task {
  id: number;
  title: string;
  description?: string;
  priority: Priority; // 1=LOW, 2=MID, 3=HIGH
  created_at: string; // ISO string
  updated_at: string; // ISO string
  deadline?: string; // ISO string or undefined
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority: Priority;
  deadline?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: Priority;
  deadline?: string;
}

export type TaskSortBy =
  | "priority"
  | "created_at"
  | "updated_at"
  | "title"
  | "id";
export interface TaskListParams {
  limit?: number;
  offset?: number;
  sort_by?: TaskSortBy;
  sort_order?: "asc" | "desc";
}

export interface TaskListResponse {
  items: Task[];
  total: number;
  page: number;
  page_size: number;
}
